src/vs/workbench/parts/execution/electron-browser/execution.contribution.ts

```diff
@@ -16,3 +36,192 @@ if (env.isWindows) {
+				'type': 'string',
+				'description': nls.localize('terminal.external.osxExec', "Customizes which terminal application to run on OS X."),
+				'default': DEFAULT_TERMINAL_OSX,
+				'isExecutable': true
+			},
+			'terminal.external.linuxExec': {
+				'type': 'string',
+				'description': nls.localize('terminal.external.linuxExec', "Customizes which terminal to run on Linux."),
+				'default': defaultTerminalLinux,
+				'isExecutable': true
+			}
+		}
+	});
+});
+
+
+export abstract class AbstarctOpenInTerminalAction extends Action {
+	private resource: uri;
+
+	constructor(
+		id: string,
+		label: string,
+		@IWorkbenchEditorService protected editorService: IWorkbenchEditorService,
+		@IWorkspaceContextService protected contextService: IWorkspaceContextService,
+		@IHistoryService protected historyService: IHistoryService
+	) {
+		super(id, label);
+
+		this.order = 49; // Allow other actions to position before or after
+	}
+
+	public setResource(resource: uri): void {
+		this.resource = resource;
+		this.enabled = !paths.isUNC(this.resource.fsPath);
+	}
+
+	public getPathToOpen(): string {
+		let pathToOpen: string;
+
+		// Try workspace path first
+		const root = this.historyService.getLastActiveWorkspaceRoot();
+		pathToOpen = this.resource ? this.resource.fsPath : (root && root.fsPath);
+
+		// Otherwise check if we have an active file open
+		if (!pathToOpen) {
+			const file = toResource(this.editorService.getActiveEditorInput(), { supportSideBySide: true, filter: 'file' });
+			if (file) {
+				pathToOpen = paths.dirname(file.fsPath); // take parent folder of file
+			}
+		}
+
+		return pathToOpen;
+	}
+}
+
+export class OpenConsoleAction extends AbstarctOpenInTerminalAction {
+
+	public static ID = 'workbench.action.terminal.openNativeConsole';
+	public static Label = env.isWindows ? nls.localize('globalConsoleActionWin', "Open New Command Prompt") :
+		nls.localize('globalConsoleActionMacLinux', "Open New Terminal");
+	public static ScopedLabel = env.isWindows ? nls.localize('scopedConsoleActionWin', "Open in Command Prompt") :
+		nls.localize('scopedConsoleActionMacLinux', "Open in Terminal");
+
+	constructor(
+		id: string,
+		label: string,
+		@ITerminalService private terminalService: ITerminalService,
+		@IWorkbenchEditorService editorService: IWorkbenchEditorService,
+		@IWorkspaceContextService contextService: IWorkspaceContextService,
+		@IHistoryService historyService: IHistoryService
+	) {
+		super(id, label, editorService, contextService, historyService);
+	}
+
+	public run(event?: any): TPromise<any> {
+		let pathToOpen = this.getPathToOpen();
+		this.terminalService.openTerminal(pathToOpen);
+
+		return TPromise.as(null);
+	}
+}
+
+export class OpenIntegratedTerminalAction extends AbstarctOpenInTerminalAction {
+
+	public static ID = 'workbench.action.terminal.openFolderInIntegratedTerminal';
+	public static Label = nls.localize('openFolderInIntegratedTerminal', "Open in Integrated Terminal");
```

Comment:
I think we should be using the same name here - adding "Integrated" makes the context menu quite a bit wider without adding much.